File: game-engine/src/main/java/org/luxons/sevenwonders/game/resources/Production.java
package org.luxons.sevenwonders.game.resources;

import java.util.Arrays;
import java.util.EnumSet;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class Production {

    private final Resources fixedResources = new Resources();

    private final Set<Set<ResourceType>> alternativeResources = new HashSet<>();

    public void addFixedResource(ResourceType type, int quantity) {
        fixedResources.add(type, quantity);
    }

    public void addChoice(ResourceType... options) {
        EnumSet<ResourceType> optionSet = EnumSet.copyOf(Arrays.asList(options));
        alternativeResources.add(optionSet);
    }

    public void addAll(Resources resources) {
        fixedResources.addAll(resources);
    }

    public void addAll(Production production) {
        fixedResources.addAll(production.getFixedResources());
        alternativeResources.addAll(production.getAlternativeResources());
    }

    public Resources getFixedResources() {
        return fixedResources;
    }

    public Set<Set<ResourceType>> getAlternativeResources() {
        return alternativeResources;
    }

    Set<Set<ResourceType>> asChoices() {
        Set<Set<ResourceType>> result = new HashSet<>(fixedResources.size() + alternativeResources.size());
        fixedResources.asList().stream().map(EnumSet::of).forEach(result::add);
        result.addAll(alternativeResources);
        return result;
    }

    public boolean contains(Resources resources) {
        if (fixedResources.contains(resources)) {
            return true;
        }
        Resources remaining = resources.minus(fixedResources);
        return containedInAlternatives(remaining);
    }

    private boolean containedInAlternatives(Resources resources) {
        return containedInAlternatives(resources, alternativeResources);
    }

    private static boolean containedInAlternatives(Resources resources, Set<Set<ResourceType>> alternatives) {
        if (resources.isEmpty()) {
            return true;
        }
        for (ResourceType type : ResourceType.values()) {
            if (resources.getQuantity(type) <= 0) {
                continue;
            }
            Set<ResourceType> candidate = findFirstAlternativeContaining(alternatives, type);
            if (candidate == null) {
                return false; // no alternative produces the resource of this entry
            }
            resources.remove(type, 1);
            alternatives.remove(candidate);
            boolean remainingAreContainedToo = containedInAlternatives(resources, alternatives);
            resources.add(type, 1);
            alternatives.add(candidate);
            if (remainingAreContainedToo) {
                return true;
            }
        }
        return false;
    }

    private static Set<ResourceType> findFirstAlternativeContaining(Set<Set<ResourceType>> alternatives,
            ResourceType type) {
        return alternatives.stream().filter(a -> a.contains(type)).findAny().orElse(null);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Production that = (Production) o;
        return Objects.equals(fixedResources, that.fixedResources) && Objects.equals(alternativeResources,
                that.alternativeResources);
    }

    @Override
    public int hashCode() {
        return Objects.hash(fixedResources, alternativeResources);
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/resources/Production.kt
package org.luxons.sevenwonders.game.resources

import java.util.Arrays
import java.util.EnumSet

class Production internal constructor() {

    val fixedResources = Resources()
    private val alternativeResources: MutableSet<Set<ResourceType>> = mutableSetOf()

    fun getAlternativeResources(): Set<Set<ResourceType>> {
        return alternativeResources
    }

    fun addFixedResource(type: ResourceType, quantity: Int) {
        fixedResources.add(type, quantity)
    }

    fun addChoice(vararg options: ResourceType) {
        val optionSet = EnumSet.copyOf(Arrays.asList(*options))
        alternativeResources.add(optionSet)
    }

    fun addAll(resources: Resources) {
        fixedResources.addAll(resources)
    }

    fun addAll(production: Production) {
        fixedResources.addAll(production.fixedResources)
        alternativeResources.addAll(production.getAlternativeResources())
    }

    internal fun asChoices(): Set<Set<ResourceType>> {
        val fixedAsChoices = fixedResources.asList().map{ EnumSet.of(it) }.toSet()
        return fixedAsChoices + alternativeResources
    }

    operator fun contains(resources: Resources): Boolean {
        if (fixedResources.containsAll(resources)) {
            return true
        }
        return containedInAlternatives(resources - fixedResources)
    }

    private fun containedInAlternatives(resources: Resources): Boolean {
        return containedInAlternatives(resources, alternativeResources)
    }

    private fun containedInAlternatives(resources: Resources, alternatives: MutableSet<Set<ResourceType>>): Boolean {
        if (resources.isEmpty) {
            return true
        }
        for (type in ResourceType.values()) {
            if (resources.getQuantity(type) <= 0) {
                continue
            }
            val candidate = findFirstAlternativeContaining(alternatives, type)
                    ?: return false // no alternative produces the resource of this entry
            resources.remove(type, 1)
            alternatives.remove(candidate)
            val remainingAreContainedToo = containedInAlternatives(resources, alternatives)
            resources.add(type, 1)
            alternatives.add(candidate)
            if (remainingAreContainedToo) {
                return true
            }
        }
        return false
    }

    private fun findFirstAlternativeContaining(
        alternatives: Set<Set<ResourceType>>,
        type: ResourceType
    ): Set<ResourceType>? {
        return alternatives.stream().filter { a -> a.contains(type) }.findAny().orElse(null)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Production

        if (fixedResources != other.fixedResources) return false
        if (alternativeResources != other.alternativeResources) return false

        return true
    }

    override fun hashCode(): Int {
        var result = fixedResources.hashCode()
        result = 31 * result + alternativeResources.hashCode()
        return result
    }
}
